pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -943,14 +1080,26 @@ func (c *Controller) AdditionalPodSubscriptions(
 	// The client wouldn't be explicitly subscribed to Pod1, so it would normally ignore it.
 	// Since it is a part of VIP1 which we are subscribe to, add it to the subscriptions
 	for s := range allAddresses {
-		for _, wl := range c.ambientIndex.Lookup(s.Name) {
+		cNetwork := c.Network(s.Name, make(labels.Instance, 0)).String()
+		for _, wl := range c.ambientIndex.Lookup(cNetwork + "/" + s.Name) {
 			// We may have gotten an update for Pod, but are subscribe to a Service.
 			// We need to force a subscription on the Pod as well
-			for addr := range wl.VirtualIps {
-				t := types.NamespacedName{Name: addr}
-				if currentSubs.Contains(t) {
-					shouldSubscribe.Insert(types.NamespacedName{Name: wl.ResourceName()})
-					break
+			switch addr := wl.Address.Type.(type) {
+			case *workloadapi.Address_Workload:
+				for vip := range addr.Workload.VirtualIps {
+					t := types.NamespacedName{Name: vip}
+					if currentSubs.Contains(t) {
+						shouldSubscribe.Insert(types.NamespacedName{Name: wl.ResourceName()})
+						break
+					}
+				}
+			case *workloadapi.Address_Service:
+				for _, networkAddress := range addr.Service.Addresses {
+					t := types.NamespacedName{Name: string(networkAddress.Address)}
```

Comment:
it looks like when we pass in `currentSubs` it's always populated with xds resource name, I think we should be using that?

```golang
	subs := sets.New(w.ResourceNames...)
	typedSubs := sets.NewWithLength[types.NamespacedName](subs.Len())
	for s := range subs {
		typedSubs.Insert(types.NamespacedName{Name: s})
	}
```